src/ui/public/courier/courier.js

```diff
@@ -113,9 +110,7 @@ uiModules.get('kibana/courier')
       }
     });
 
-    const onFatalDefer = Promise.defer();
-    onFatalDefer.promise.then(self.close);
-    Notifier.fatalCallbacks.push(onFatalDefer.resolve);
+    Notifier.fatalCallbacks.push(() => self.close());
```

Comment:
I don't think this will be easy to add a test for.  I tried, but the problem is that calling notify.fatal will actually throw the fatal screen and interrupt the tests.  If I mock it out, then I'm not really testing that the callbacks are being called when fatal is called.

Reply:
Completely agree regarding the automated test. I intended to ask for an easy way for me to simulate this code-path being hit manually, just to ensure we didn't unintentionally cause a regression. When trying to determine the history of this syntax, I came across this [commit](https://github.com/elastic/kibana/commit/ffe92528f4a1ded00b7773bd61668c812c956229) with the message of `[courier] move fatal callback into Courier to magically fix issue (thank git bisect)` that sounds quite ominous.

Reply:
I dug a bit deeper and there are some interesting differences, though I don't think any of them were intentional (any chance you can weigh in here @spalger?)

Difference #1 is that using resolve makes it only call self.close once.  I can fix this via:

```
    const closeOnFatal = _.once(self.close);
    Notifier.fatalCallbacks.push(closeOnFatal);
```

Difference number two is that when the callbacks are called, the order will change - if it's a resolve self.close gets called asynchronously, if a function it gets called synchronously.

I'm not sure if this is due to the async/sync difference above, but hitting the `throw new Error('Aborting all pending requests failed.');` line will show up in different ways.  In the old way it's an additional error on the fatal page (in addition to hitting a few extra errors).  In the new way, it ends up showing up as an error banner and the fatal page goes away.  But we are talking about a use case that just hit a fatal error and another error, so it doesn't seem like something worth keeping the old code around for.

Lets see if @spalger can weigh in to give us some confidence.  I find the defer syntax to be really confusing and using it will make it harder to switch to native promises.

Reply:
I'm pretty confident that the use of a promise here was to ensure that the digest loop was maintained or created if the fatal error occurred outside of a digest loop... but it's been so long...

Closing the courier aborts all requests, which triggers other logic, and since the Notifier isn't angular specific it could be triggering that tree of logic outside of a digest cycle. I'm not concerned about self.close being called multiple times, but perhaps we should use `$rootScope.$evalAsync()` within the fatal callback to ensure it all happens in a digest cycle... 

Sorry I didn't keep good records about what bug I was fixing in ffe92528f4a1ded00b7773bd61668c812c956229

Reply:
If we aren't *sure* we need the evalAsync function, I'm hesitant to add it in.  I can trigger a fatal error from inside react, outside of a digest cycle (I think, judging by the call stack in the browser) and close gets called as expected, and the fatal page shows up.

I just don't want to throw code in with the hopes of maybe catching some errors. I'd rather wait until a legit issue pops up, we can then write a test for it, and figure out the best way to fix it.  

Eventually I think courier should be angular free as well.